dev-tools/releng/src/job.rs

```diff
@@ -0,0 +1,284 @@
+    wait_for: Vec<oneshot::Receiver<()>>,
+    notify: Vec<oneshot::Sender<()>>,
+}
+
+pub(crate) struct Selector<'a> {
+    jobs: &'a mut Jobs,
+    name: String,
+}
+
+impl Jobs {
+    pub(crate) fn new(
+        logger: &Logger,
+        permits: Arc<Semaphore>,
+        log_dir: &Utf8Path,
+    ) -> Jobs {
+        Jobs {
+            logger: logger.clone(),
+            permits,
+            log_dir: log_dir.to_owned(),
+            map: HashMap::new(),
+        }
+    }
+
+    pub(crate) fn push<F>(
+        &mut self,
+        name: impl AsRef<str>,
+        future: F,
+    ) -> Selector<'_>
+    where
+        F: Future<Output = Result<()>> + 'static,
+    {
+        let name = name.as_ref().to_owned();
+        assert!(!self.map.contains_key(&name), "duplicate job name {}", name);
+        self.map.insert(
+            name.clone(),
+            Job {
+                future: Box::pin(run_job(
+                    self.logger.clone(),
+                    self.permits.clone(),
+                    name.clone(),
+                    future,
+                )),
+                wait_for: Vec::new(),
+                notify: Vec::new(),
+            },
+        );
+        Selector { jobs: self, name }
+    }
+
+    pub(crate) fn push_command(
+        &mut self,
+        name: impl AsRef<str>,
+        command: &mut Command,
+    ) -> Selector<'_> {
+        let name = name.as_ref().to_owned();
+        assert!(!self.map.contains_key(&name), "duplicate job name {}", name);
+        self.map.insert(
+            name.clone(),
+            Job {
+                future: Box::pin(spawn_with_output(
+                    // terrible hack to deal with the `Command` builder
+                    // returning &mut
+                    std::mem::replace(command, Command::new("false")),
+                    self.logger.clone(),
+                    self.permits.clone(),
+                    name.clone(),
+                    self.log_dir.join(&name).with_extension("log"),
+                )),
+                wait_for: Vec::new(),
+                notify: Vec::new(),
+            },
+        );
+        Selector { jobs: self, name }
+    }
+
+    pub(crate) fn select(&mut self, name: impl AsRef<str>) -> Selector<'_> {
+        Selector { jobs: self, name: name.as_ref().to_owned() }
+    }
+
+    pub(crate) async fn run_all(self) -> Result<()> {
+        self.map
+            .into_values()
+            .map(Job::run)
+            .collect::<FuturesUnordered<_>>()
+            .try_collect::<()>()
+            .await
+    }
+}
+
+impl Job {
+    async fn run(self) -> Result<()> {
+        let result: Result<(), RecvError> = self
+            .wait_for
+            .into_iter()
+            .collect::<FuturesUnordered<_>>()
+            .try_collect::<()>()
+            .await;
+        result.map_err(|_| anyhow!("dependency failed"))?;
+
+        self.future.await?;
+        for sender in self.notify {
+            sender.send(()).ok();
```

Comment:
Can we log the situation where the receiver is closed instead of ignoring it?

Reply:
I think the only reason a receiver would close is if the tasks are all cancelled, which would happen if any task fails or if the user hits Ctrl-C. Assuming that's the case, this would only log in a situation where it would be unnecessarily noisy (we already log why a task failed or bubble the error all the way back up).

Reply:
Ah ok -- that seems reasonable then.